### ISSUE TITLE
[IJ/AS plugin] Ignore id field in Unused field inspection

### ISSUE THREAD

**Issue by sonatard** [apollographql/apollo-kotlin#5172]
### Use case

I want to ignore the id field in Unused field inspection.
This is because the id field is retrieved for cache updates, but it is very often not accessed from Kotlin.

### Describe the solution you'd like

id: ID is ignore, Specify the field name to ignore or Specify the type to ignore

**Comment by BoD** [comment]
Hi! 

That's a good point, I think there might be a lot of cases where you need to select fields only for the cache to be up to date, but they won't be used in the Kotlin code and thus highlighted. This shouldn't be specific to id fields though?

You can suppress the inspection with `# noinspection ApolloUnusedField` on top of an operation / fragment. Unfortunately at the moment this doesn't work on individual fields though.

Maybe can you elaborate with your concrete example? Thanks!

**Comment by sonatard** [comment]
> This shouldn't be specific to id fields though?
Yes. The mutation response is often not used directly by Kotlin, and is mostly for updating the cache.

> Maybe can you elaborate with your concrete example? Thanks!

The IDs such as lines.nodes.id, productVariant.id, and product.id are not referenced from Kotlin.

```graphql
query StorefrontAddCartLine($cartID: ID!, $countryCode: CountryCode!, $languageCode: LanguageCode!) @inContext(country: $countryCode, language: $languageCode) {
    cart(id: $cartID) {
        id
        totalQuantity
        lines(first: 100) {
            nodes {
                id # only for cache
                quantity
                merchandise {
                    ... on ProductVariant {
                        id # only for cache
                        product {
                            id # only for cache
                            tags
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by BoD** [comment]
Are you [declaring](https://www.apollographql.com/docs/kotlin/caching/declarative-ids) your id fields with `@typePolicy`? If yes you should be able to omit the `id` fields in your selections, they will still automatically be selected under the hood.

**Comment by sonatard** [comment]
I didn't know that the cache would be updated even if the id field was omitted. It's wonderful news!

Can the id field be omitted in the same way with Programmatic cache IDs? Defining Declarative cache IDs for many types is cumbersome, so I would like to achieve this with Programmatic cache IDs if possible.

**Comment by BoD** [comment]
The automatic addition of id fields to selections is done during the codegen, and works by finding the `@typePolicy` directives - so that won't work with programmatic cache IDs which is done at runtime. I _think_ it could be done in theory but there is nothing like that in the library at the moment.

**Comment by BoD** [comment]
In any case, we could certainly add a setting for this inspection, with a list of names or regexp of fields that should be ignored 👍

**Comment by sonatard** [comment]
Thanks! I try set `extend interface Node @typePolicy(keyFields: "id")`
If the title of this issue should be changed to something more appropriate, please make the change.

**Comment by BoD** [comment]
Fixed via #5195 and #5197
